Question answering with Transformers example added

merveenoyan opened the pull request:
Hello 👋 
I wanted to add an example script on Question Answering using 🤗 Transformers but I ran across an issue with autogeneration of notebooks from script. (I formatted the script and there's nothing wrong with the style, style tests pass)

See [this issue I opened](https://github.com/keras-team/keras-io/issues/774) about it. Maybe I am mistaken in a step. Opening this as a draft PR since I couldn't autogenerate the notebook.

cc'ing here: @fchollet (because I don't know anyone else I can ask from 😔)

fchollet commented on examples/nlp/question-answering.py:
```diff
@@ -0,0 +1,422 @@
+Description: Question answering implementation using Keras and 🤗Transformers.
+"""
+"""shell
+!pip install git+https://github.com/huggingface/transformers.git
```
Remove `!` characters, they get added automatically

fchollet commented on examples/nlp/question-answering.py:
```diff
@@ -0,0 +1,422 @@
+"""
+
+"""
+# Fine-tuning a model on a question answering task
```
`#` is only for the title, all section titles should use `##`

fchollet commented on examples/nlp/question-answering.py:
```diff
@@ -0,0 +1,422 @@
+your task of interest?
+
+In this example, we'll show you how to load a model from the Hugging Face [🤗
+Transformers](https://github.com/huggingface/transformers) library to tackle this
```
Make sure to follow the style guide with regard to things like markdown link formatting, etc. https://github.com/keras-team/keras-io/blob/master/contributor_guide.md

fchollet commented on examples/nlp/question-answering.py:
```diff
@@ -0,0 +1,422 @@
+[here](https://huggingface.co/spaces/keras-io/keras-qa).
+"""
+
+model.push_to_hub("transformers-qa", organization="keras-io")
```
It seems questionable to push the model every time someones runs the script -- this should be kept as a code block in the text rather than as executable code

fchollet commented on examples/nlp/question-answering.py:
```diff
@@ -0,0 +1,422 @@
+"""
+Title: Question Answering with Transformers
```
"HuggingFace Transformers"

fchollet commented on examples/nlp/question-answering.py:
```diff
@@ -0,0 +1,422 @@
+mentioned above. Let's take a look at what a single training example looks like.
+"""
+
+datasets["train"][0]
```
The script should also work as a standalone Python script, so use `print()` to display things

fchollet commented on examples/nlp/question-answering.py:
```diff
@@ -0,0 +1,422 @@
+"""
+
+tokenized_datasets = datasets.map(
+    prepare_train_features, batched=True, remove_columns=datasets["train"].column_names
```
Make sure to use `num_parallel_calls` (you can default to `tf.data.AUTOTUNE`)

merveenoyan replied:
I used `num_proc` which is `datasets` equivalent of `num_parallel_calls`.

fchollet reviewed (commented):
Thank you for the PR! 👍

merveenoyan commented:
I autogeneration script trains the model, so I'm currently training the model (takes a while) and it will be ready to be reviewed. Thanks a lot 🤗

fchollet commented:
>  autogeneration script trains the model, so I'm currently training the model (takes a while) and it will be ready to be reviewed.

Pease don't spend too much running the autogeneration script, since it will have to be re-run when the Python file is finalized! How long does it take to run currently?

merveenoyan commented:
Before model it doesn't take too long to be honest, model takes around 30 hours on CPU. Should I just submit?

fchollet commented:
> model takes around 30 mins on GPU to train

Thanks for the answer -- Ok, that's a bit heavy but acceptable.

Is the PR ready to review? If so please remove the draft tag and I'll do the review -- also you'll need to sign the CLA at https://cla.developers.google.com/clas thank you!

merveenoyan commented:
@fchollet Hello 👋🏻 @julien-c signed on behalf of Hugging Face, can you sign on your end as well?

fchollet commented on examples/nlp/question-answering.py:
```diff
@@ -0,0 +1,427 @@
+Last modified: 13/01/2022
+Description: Question answering implementation using Keras and Hugging Face Transformers.
+"""
+"""shell
```
Please add an "introduction" section for the problem description and the installation instructions

merveenoyan replied:
Added in my last commit 🙌🏼

fchollet reviewed (changes requested):
Thank you for the update! Please note that I pushed some copyedits: please pull them first.

merveenoyan commented:
Added an introduction section on top of the notebook

fchollet commented on examples/nlp/question-answering.py:
```diff
@@ -0,0 +1,426 @@
+representing the start and end of the span containing the answer. This variant is called
+"extractive question answering".
+
+## Fine-tuning a model on a question answering task
```
You can remove this title, this can be a single introduction block. If you need you can turn the title into a single introduction sentence.

fchollet commented on examples/nlp/question-answering.py:
```diff
@@ -0,0 +1,426 @@
+Description: Question answering implementation using Keras and Hugging Face Transformers.
+"""
+"""
+# Introduction to Question Answering
```
All sections should use `##`, `#` is only for the main title (automatically added)

fchollet commented on examples/nlp/question-answering.py:
```diff
@@ -0,0 +1,426 @@
+A list of possible answers are supplied with each question, and the model simply needs to
+return a probability distribution over the options. A more challenging variant of
+question answering, which is more applicable to real-life tasks, is when the options are
+not provided. Instead, the model is given an input document -called context- and a 
```
Use `--` for em-dashes, surrounded by spaces

fchollet reviewed (commented):
Thanks for the update!

fchollet reviewed (approved):
LGTM, thank you! Please add the generated files.

merveenoyan commented:
@fchollet Hello 👋 
I generated the notebook, and training took me less, I trained for 20 min or so for one epoch so the script didn't raise an error 🤩😅

I changed the name of the script from kebab-case to snake_case for consistency, generated the files and added them. My only problem is the header appearing twice, though there's no such thing in script. (see below)
<img width="1107" alt="Ekran Resmi 2022-01-25 10 46 53" src="https://user-images.githubusercontent.com/53175384/150952696-0417f4f4-548e-4bba-89fc-af432ef13c23.png">
<img width="788" alt="Ekran Resmi 2022-01-25 10 47 16" src="https://user-images.githubusercontent.com/53175384/150952758-d00156ec-9c31-4465-a07c-f77255f2fbd0.png">

Also, would you like me to get rid of the cells of the outputs that look a little verbose? 

<img width="1051" alt="Ekran Resmi 2022-01-25 10 48 17" src="https://user-images.githubusercontent.com/53175384/150952952-89b5896b-e6ca-49bf-9aaa-5fe57a8128dc.png">

fchollet commented:
> I changed the name of the script from kebab-case to snake_case for consistency, generated the files and added them. My only problem is the header appearing twice, though there's no such thing in script. (see below)

This is because there's a leading space in the md file, you can remove it

> Also, would you like me to get rid of the cells of the outputs that look a little verbose?

Sure, you can manually remove unnecessary logs from the md file.

Thank you!

merveenoyan commented:
@fchollet I think we're good to go! 🤩  thanks a lot for the never-ending support 🙌🏼

<img width="1048" alt="Ekran Resmi 2022-01-26 10 07 50" src="https://user-images.githubusercontent.com/53175384/151134524-ff1a0f15-c4d2-4eef-af54-e817c9c3f299.png">
Also removed two verbose output blocks from markdown file.

fchollet commented:
Awesome! Thanks again for the great contribution 👍
